File: Autonomous_Mobile_Robots_Course/Chapter7/Lesson3/Chapter7_Lesson3.py
#!/usr/bin/env python3
"""
Chapter7_Lesson3.py
Consistency and Linearization Pitfalls — EKF vs UKF on a simple AMR localization example.

Dependencies:
  - numpy
  - scipy (for chi-square bounds)

This script:
  1) Simulates a unicycle-like planar robot (x, y, theta) with process noise.
  2) Uses range-bearing measurements to known landmarks.
  3) Runs EKF and UKF.
  4) Reports ANEES and ANIS consistency statistics with 95% bounds.

Author: (course material)
"""
import math
import numpy as np

try:
    from scipy.stats import chi2
except Exception as e:
    chi2 = None


def wrap_angle(a: float) -> float:
    """Wrap angle to (-pi, pi]."""
    return (a + math.pi) % (2.0 * math.pi) - math.pi


def f_motion(x: np.ndarray, u: np.ndarray, dt: float) -> np.ndarray:
    """Discrete-time motion model (no noise inside; noise added externally)."""
    px, py, th = x
    v, om = u
    nx = px + v * dt * math.cos(th)
    ny = py + v * dt * math.sin(th)
    nth = wrap_angle(th + om * dt)
    return np.array([nx, ny, nth], dtype=float)


def jacobian_F(x: np.ndarray, u: np.ndarray, dt: float) -> np.ndarray:
    """Jacobian of f w.r.t state x at (x,u)."""
    _, _, th = x
    v, _ = u
    F = np.eye(3)
    F[0, 2] = -v * dt * math.sin(th)
    F[1, 2] =  v * dt * math.cos(th)
    return F


def h_meas(x: np.ndarray, landmark: np.ndarray) -> np.ndarray:
    """Range-bearing measurement to a landmark."""
    px, py, th = x
    dx = landmark[0] - px
    dy = landmark[1] - py
    r = math.hypot(dx, dy)
    b = wrap_angle(math.atan2(dy, dx) - th)
    return np.array([r, b], dtype=float)


def jacobian_H(x: np.ndarray, landmark: np.ndarray) -> np.ndarray:
    """Jacobian of h w.r.t state x."""
    px, py, th = x
    dx = landmark[0] - px
    dy = landmark[1] - py
    r2 = dx*dx + dy*dy
    r = math.sqrt(r2) if r2 > 1e-12 else 1e-6
    H = np.zeros((2, 3))
    H[0, 0] = -dx / r
    H[0, 1] = -dy / r
    H[0, 2] = 0.0
    H[1, 0] =  dy / r2
    H[1, 1] = -dx / r2
    H[1, 2] = -1.0
    return H


def ekf_step(x: np.ndarray, P: np.ndarray, u: np.ndarray, z: np.ndarray,
             dt: float, Q: np.ndarray, R: np.ndarray, landmark: np.ndarray):
    """One EKF predict + update."""
    # Predict
    F = jacobian_F(x, u, dt)
    x_pred = f_motion(x, u, dt)
    P_pred = F @ P @ F.T + Q

    # Update
    H = jacobian_H(x_pred, landmark)
    z_pred = h_meas(x_pred, landmark)
    nu = z - z_pred
    nu[1] = wrap_angle(nu[1])  # bearing innovation
    S = H @ P_pred @ H.T + R
    K = P_pred @ H.T @ np.linalg.inv(S)

    x_upd = x_pred + K @ nu
    x_upd[2] = wrap_angle(x_upd[2])
    I = np.eye(3)
    P_upd = (I - K @ H) @ P_pred @ (I - K @ H).T + K @ R @ K.T  # Joseph form

    return x_upd, P_upd, nu, S


def sigma_points(x: np.ndarray, P: np.ndarray, alpha=1e-3, beta=2.0, kappa=0.0):
    """Scaled unscented sigma points."""
    n = x.size
    lam = alpha*alpha*(n + kappa) - n
    S = np.linalg.cholesky((n + lam) * P)

    X = np.zeros((2*n + 1, n))
    X[0] = x
    for i in range(n):
        X[1 + i]     = x + S[:, i]
        X[1 + n + i] = x - S[:, i]

    Wm = np.full(2*n + 1, 1.0 / (2.0*(n + lam)))
    Wc = np.full(2*n + 1, 1.0 / (2.0*(n + lam)))
    Wm[0] = lam / (n + lam)
    Wc[0] = lam / (n + lam) + (1.0 - alpha*alpha + beta)

    return X, Wm, Wc


def mean_angle(weights: np.ndarray, angles: np.ndarray) -> float:
    s = np.sum(weights * np.sin(angles))
    c = np.sum(weights * np.cos(angles))
    return math.atan2(s, c)


def ukf_step(x: np.ndarray, P: np.ndarray, u: np.ndarray, z: np.ndarray,
             dt: float, Q: np.ndarray, R: np.ndarray, landmark: np.ndarray):
    """One UKF predict + update (state dimension 3, measurement dimension 2)."""
    n = x.size
    m = z.size

    # Predict
    X, Wm, Wc = sigma_points(x, P)
    X_pred = np.zeros_like(X)
    for i in range(X.shape[0]):
        X_pred[i] = f_motion(X[i], u, dt)

    x_pred = np.zeros(n)
    x_pred[0] = np.sum(Wm * X_pred[:, 0])
    x_pred[1] = np.sum(Wm * X_pred[:, 1])
    x_pred[2] = mean_angle(Wm, X_pred[:, 2])

    P_pred = np.zeros((n, n))
    for i in range(X_pred.shape[0]):
        dx = X_pred[i] - x_pred
        dx[2] = wrap_angle(dx[2])
        P_pred += Wc[i] * np.outer(dx, dx)
    P_pred += Q

    # Update
    Zsig = np.zeros((X_pred.shape[0], m))
    for i in range(X_pred.shape[0]):
        Zsig[i] = h_meas(X_pred[i], landmark)

    z_pred = np.zeros(m)
    z_pred[0] = np.sum(Wm * Zsig[:, 0])
    z_pred[1] = mean_angle(Wm, Zsig[:, 1])

    S = np.zeros((m, m))
    Cxz = np.zeros((n, m))
    for i in range(Zsig.shape[0]):
        dz = Zsig[i] - z_pred
        dz[1] = wrap_angle(dz[1])
        dx = X_pred[i] - x_pred
        dx[2] = wrap_angle(dx[2])
        S += Wc[i] * np.outer(dz, dz)
        Cxz += Wc[i] * np.outer(dx, dz)
    S += R

    nu = z - z_pred
    nu[1] = wrap_angle(nu[1])
    K = Cxz @ np.linalg.inv(S)

    x_upd = x_pred + K @ nu
    x_upd[2] = wrap_angle(x_upd[2])
    P_upd = P_pred - K @ S @ K.T

    return x_upd, P_upd, nu, S


def chi2_bounds_for_average(dof: int, N: int, alpha: float = 0.05):
    """Bounds for average of chi-square/dof statistic: (N * avg) ~ chi2(N*dof)."""
    if chi2 is None:
        return None
    lo = chi2.ppf(alpha/2.0, N*dof) / N
    hi = chi2.ppf(1.0 - alpha/2.0, N*dof) / N
    return lo, hi


def run_mc(filter_step, dt: float, steps: int, trials: int, seed: int = 0):
    rng = np.random.default_rng(seed)

    # Landmarks (known map)
    landmarks = np.array([[5.0,  0.0],
                          [0.0,  5.0],
                          [5.0,  5.0]], dtype=float)

    # Noise settings (intentionally a bit "mismatched" to amplify linearization effects)
    Q = np.diag([0.08**2, 0.08**2, (3.0*math.pi/180.0)**2])  # process noise covariance
    R = np.diag([0.15**2, (2.0*math.pi/180.0)**2])          # measurement noise covariance

    nees_list = []
    nis_list = []

    for tr in range(trials):
        # True initial state (unknown to filter)
        x_true = np.array([0.0, 0.0, 0.0], dtype=float)

        # Filter initial belief (purposely slightly biased)
        x_hat = np.array([0.5, -0.4, 10.0*math.pi/180.0], dtype=float)
        P = np.diag([0.8**2, 0.8**2, (15.0*math.pi/180.0)**2])

        for k in range(steps):
            # Control: forward + turning (induces nonlinearity)
            v = 1.0 + 0.2*math.sin(0.05*k)
            om = 0.35 + 0.25*math.sin(0.03*k)

            u = np.array([v, om], dtype=float)

            # Simulate true process with noise
            w = rng.multivariate_normal(np.zeros(3), Q)
            x_true = f_motion(x_true, u, dt) + w
            x_true[2] = wrap_angle(x_true[2])

            # Choose a landmark based on time (switching adds mild nonlinearity)
            lm = landmarks[k % landmarks.shape[0]]

            # Measurement
            z = h_meas(x_true, lm)
            v_meas = rng.multivariate_normal(np.zeros(2), R)
            z = z + v_meas
            z[1] = wrap_angle(z[1])

            # Filter step
            x_hat, P, nu, S = filter_step(x_hat, P, u, z, dt, Q, R, lm)

            # Consistency stats
            e = x_true - x_hat
            e[2] = wrap_angle(e[2])

            nees = float(e.T @ np.linalg.inv(P) @ e)
            nis  = float(nu.T @ np.linalg.inv(S) @ nu)
            nees_list.append(nees)
            nis_list.append(nis)

    nees_arr = np.array(nees_list)
    nis_arr = np.array(nis_list)
    return nees_arr.mean(), nis_arr.mean(), 3, 2, trials*steps


def main():
    # Two dt values: small dt often improves linearization; large dt worsens it
    configs = [
        ("dt=0.05 (mild nonlinearity)", 0.05),
        ("dt=0.20 (stronger nonlinearity)", 0.20),
    ]
    steps = 200
    trials = 40
    alpha = 0.05

    for label, dt in configs:
        print("\n=== ", label, " ===")
        ekf_anees, ekf_anis, n, m, N = run_mc(ekf_step, dt, steps, trials, seed=1)
        ukf_anees, ukf_anis, _, _, _ = run_mc(ukf_step, dt, steps, trials, seed=1)

        print(f"EKF  ANEES={ekf_anees:.3f}   ANIS={ekf_anis:.3f}")
        print(f"UKF  ANEES={ukf_anees:.3f}   ANIS={ukf_anis:.3f}")

        if chi2 is not None:
            nees_lo, nees_hi = chi2_bounds_for_average(n, N, alpha)
            nis_lo, nis_hi = chi2_bounds_for_average(m, N, alpha)
            print(f"95% bounds for ANEES (dof={n}): [{nees_lo:.3f}, {nees_hi:.3f}]")
            print(f"95% bounds for ANIS  (dof={m}): [{nis_lo:.3f}, {nis_hi:.3f}]")
        else:
            print("SciPy not available: chi-square bounds not computed.")

        print("Interpretation: values systematically above bounds indicate overconfidence (inconsistency).")

if __name__ == "__main__":
    main()

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter7/Lesson3/Chapter7_Lesson3.cpp
// Chapter7_Lesson3.cpp
// Consistency and Linearization Pitfalls — EKF + NEES/NIS on a 2D AMR localization example.
//
// Build (example):
//   g++ -O2 -std=c++17 Chapter7_Lesson3.cpp -I /usr/include/eigen3 -o lesson7_3
//
// Optional bounds (requires Boost):
//   (most Linux distros already provide boost headers)
//
// This program prints ANEES/ANIS and 95% bounds (if Boost is available).

#include <cmath>
#include <iostream>
#include <random>
#include <vector>

#include <Eigen/Dense>

#ifdef __has_include
#  if __has_include(<boost/math/distributions/chi_squared.hpp>)
#    include <boost/math/distributions/chi_squared.hpp>
#    define HAS_BOOST_CHI2 1
#  endif
#endif

static double wrapAngle(double a) {
    a = std::fmod(a + M_PI, 2.0 * M_PI);
    if (a < 0) a += 2.0 * M_PI;
    return a - M_PI;
}

static Eigen::Vector3d fMotion(const Eigen::Vector3d& x, const Eigen::Vector2d& u, double dt) {
    const double px = x(0), py = x(1), th = x(2);
    const double v = u(0), om = u(1);
    Eigen::Vector3d xn;
    xn(0) = px + v * dt * std::cos(th);
    xn(1) = py + v * dt * std::sin(th);
    xn(2) = wrapAngle(th + om * dt);
    return xn;
}

static Eigen::Matrix3d jacF(const Eigen::Vector3d& x, const Eigen::Vector2d& u, double dt) {
    const double th = x(2);
    const double v  = u(0);
    Eigen::Matrix3d F = Eigen::Matrix3d::Identity();
    F(0,2) = -v * dt * std::sin(th);
    F(1,2) =  v * dt * std::cos(th);
    return F;
}

static Eigen::Vector2d hMeas(const Eigen::Vector3d& x, const Eigen::Vector2d& lm) {
    const double px = x(0), py = x(1), th = x(2);
    const double dx = lm(0) - px;
    const double dy = lm(1) - py;
    const double r  = std::hypot(dx, dy);
    const double b  = wrapAngle(std::atan2(dy, dx) - th);
    return Eigen::Vector2d(r, b);
}

static Eigen::Matrix<double,2,3> jacH(const Eigen::Vector3d& x, const Eigen::Vector2d& lm) {
    const double px = x(0), py = x(1);
    const double dx = lm(0) - px;
    const double dy = lm(1) - py;
    const double r2 = dx*dx + dy*dy;
    const double r  = std::sqrt(std::max(r2, 1e-12));

    Eigen::Matrix<double,2,3> H;
    H.setZero();
    H(0,0) = -dx / r;
    H(0,1) = -dy / r;
    H(1,0) =  dy / r2;
    H(1,1) = -dx / r2;
    H(1,2) = -1.0;
    return H;
}

struct Stats {
    double ane_es = 0.0;
    double an_is  = 0.0;
    int    dof_x  = 3;
    int    dof_z  = 2;
    int    N      = 0;
};

static Stats runMC(double dt, int steps, int trials, unsigned seed) {
    std::mt19937 rng(seed);
    std::normal_distribution<double> stdn(0.0, 1.0);

    Eigen::Matrix3d Q = Eigen::Matrix3d::Zero();
    Q(0,0) = 0.08*0.08;
    Q(1,1) = 0.08*0.08;
    Q(2,2) = std::pow(3.0*M_PI/180.0, 2);

    Eigen::Matrix2d R = Eigen::Matrix2d::Zero();
    R(0,0) = 0.15*0.15;
    R(1,1) = std::pow(2.0*M_PI/180.0, 2);

    auto sampleGaussian = [&](const Eigen::MatrixXd& Cov) {
        Eigen::LLT<Eigen::MatrixXd> llt(Cov);
        Eigen::MatrixXd L = llt.matrixL();
        Eigen::VectorXd z(Cov.rows());
        for (int i=0;i<z.size();++i) z(i) = stdn(rng);
        return L * z;
    };

    std::vector<Eigen::Vector2d> lms = {
        Eigen::Vector2d(5.0, 0.0),
        Eigen::Vector2d(0.0, 5.0),
        Eigen::Vector2d(5.0, 5.0)
    };

    double nees_sum = 0.0;
    double nis_sum  = 0.0;
    int    count    = 0;

    for (int tr=0; tr<trials; ++tr) {
        Eigen::Vector3d x_true(0.0, 0.0, 0.0);
        Eigen::Vector3d x_hat (0.5, -0.4, 10.0*M_PI/180.0);
        Eigen::Matrix3d P = Eigen::Matrix3d::Zero();
        P(0,0) = 0.8*0.8;
        P(1,1) = 0.8*0.8;
        P(2,2) = std::pow(15.0*M_PI/180.0, 2);

        for (int k=0; k<steps; ++k) {
            const double v  = 1.0 + 0.2*std::sin(0.05*k);
            const double om = 0.35 + 0.25*std::sin(0.03*k);
            Eigen::Vector2d u(v, om);

            // True propagation with noise
            Eigen::Vector3d w = sampleGaussian(Q);
            x_true = fMotion(x_true, u, dt) + w;
            x_true(2) = wrapAngle(x_true(2));

            // Landmark selection
            const Eigen::Vector2d& lm = lms[k % (int)lms.size()];

            // Measurement with noise
            Eigen::Vector2d z = hMeas(x_true, lm) + sampleGaussian(R);
            z(1) = wrapAngle(z(1));

            // EKF predict
            Eigen::Matrix3d F = jacF(x_hat, u, dt);
            Eigen::Vector3d x_pred = fMotion(x_hat, u, dt);
            Eigen::Matrix3d P_pred = F * P * F.transpose() + Q;

            // EKF update
            Eigen::Matrix<double,2,3> H = jacH(x_pred, lm);
            Eigen::Vector2d z_pred = hMeas(x_pred, lm);
            Eigen::Vector2d nu = z - z_pred;
            nu(1) = wrapAngle(nu(1));

            Eigen::Matrix2d S = H * P_pred * H.transpose() + R;
            Eigen::Matrix<double,3,2> K = P_pred * H.transpose() * S.inverse();

            x_hat = x_pred + K * nu;
            x_hat(2) = wrapAngle(x_hat(2));

            Eigen::Matrix3d I = Eigen::Matrix3d::Identity();
            // Joseph form
            P = (I - K*H) * P_pred * (I - K*H).transpose() + K*R*K.transpose();

            // Stats
            Eigen::Vector3d e = x_true - x_hat;
            e(2) = wrapAngle(e(2));

            const double nees = (e.transpose() * P.inverse() * e)(0);
            const double nis  = (nu.transpose() * S.inverse() * nu)(0);

            nees_sum += nees;
            nis_sum  += nis;
            count++;
        }
    }

    Stats st;
    st.ane_es = nees_sum / (double)count;
    st.an_is  = nis_sum  / (double)count;
    st.N      = count;
    return st;
}

#ifdef HAS_BOOST_CHI2
static std::pair<double,double> chi2BoundsAverage(int dof, int N, double alpha=0.05) {
    using boost::math::chi_squared_distribution;
    chi_squared_distribution<double> dist((double)(N*dof));
    const double lo = boost::math::quantile(dist, alpha/2.0) / (double)N;
    const double hi = boost::math::quantile(dist, 1.0 - alpha/2.0) / (double)N;
    return {lo, hi};
}
#endif

int main() {
    const int steps  = 200;
    const int trials = 40;

    for (auto dt : {0.05, 0.20}) {
        std::cout << "\n=== dt=" << dt << " ===\n";
        Stats st = runMC(dt, steps, trials, 1);

        std::cout << "EKF ANEES=" << st.ane_es << "  (expected ~ " << st.dof_x << ")\n";
        std::cout << "EKF ANIS =" << st.an_is  << "  (expected ~ " << st.dof_z << ")\n";

#ifdef HAS_BOOST_CHI2
        auto neesB = chi2BoundsAverage(st.dof_x, st.N, 0.05);
        auto nisB  = chi2BoundsAverage(st.dof_z, st.N, 0.05);
        std::cout << "95% bounds for ANEES: [" << neesB.first << ", " << neesB.second << "]\n";
        std::cout << "95% bounds for ANIS : [" << nisB.first  << ", " << nisB.second  << "]\n";
#else
        std::cout << "(Boost chi-square not detected; bounds not printed.)\n";
#endif
        std::cout << "Interpretation: ANEES or ANIS systematically above bounds indicates overconfidence.\n";
    }
    return 0;
}
